pkg/util/iptree/iptree.go

```diff
@@ -0,0 +1,592 @@
+			panic(fmt.Sprintf("wrong ancestor %s: child %s N %s", ancestor.String(), prefix.String(), n.prefix.String()))
+		}
+		link.child[idxChild] = child
+		link.child[idxN] = n
+		return false
+	}
+
+	// if already exist update it and make it public
+	if n != nil && n.prefix == prefix {
+		if n.isPublic() {
+			n.val = v
+			n.public = true
+			return true
+		}
+		n.val = v
+		n.public = true
+		return false
+	}
+
+	return false
+}
+
+// Delete the exact prefix and return true if it existed
+func (t *Tree[T]) DeletePrefix(prefix netip.Prefix) bool {
+	root := t.rootV4
+	if prefix.Addr().Is6() {
+		root = t.rootV6
+	}
+	var parent *node[T]
+	n := root
+	// bit position is given by the mask bits
+	bitPosition := 0
+	// mask the address
+	address := prefix.Masked().Addr()
+	mask := prefix.Bits()
+	for bitPosition < mask {
+		// Look for a child checking the bit position after the mask
+		parent = n
+		n = n.child[getBitFromAddr(address, bitPosition+1)]
+		if n == nil {
+			return false
+		}
+		// check we are in the right branch comparing the suffixes
+		if !n.prefix.Contains(address) {
+			return false
+		}
+		// update the new bit position with the new node mask
+		bitPosition = n.prefix.Bits()
+	}
+	// check if the node contains the prefix we want to delete
+	if n.prefix != prefix {
+		return false
+	}
+	// Delete the value
+	n.public = false
+	var zeroT T
+	n.val = zeroT
+
+	nodeChildren := 0
+	if n.child[0] != nil {
+		nodeChildren++
+	}
+	if n.child[1] != nil {
+		nodeChildren++
+	}
+	// If there is a parent and this node does not have any children
+	// this is a leaf so we can delete this node.
+	// - parent -> child(to be deleted)
+	if parent != nil && nodeChildren == 0 {
+		if parent.child[0] != nil && parent.child[0] == n {
+			parent.child[0] = nil
+		} else if parent.child[1] != nil && parent.child[1] == n {
+			parent.child[1] = nil
+		} else {
+			panic("wrong parent")
+		}
+		n = nil
+	}
+	// Check if we should merge this node
+	// The root node can not be merged
+	if n != root && nodeChildren == 1 {
+		n.mergeChild()
+	}
+	// Check if we should merge the parent's other child
+	// parent -> deletedNode
+	//        |--> child
+	parentChildren := 0
+	if parent != nil {
+		if parent.child[0] != nil {
+			parentChildren++
+		}
+		if parent.child[1] != nil {
+			parentChildren++
+		}
+		if parent != root && parentChildren == 1 && !parent.isPublic() {
+			parent.mergeChild()
+		}
+	}
+	return true
+}
+
+// for testing
+func (t *Tree[T]) len(isV6 bool) int {
+	count := 0
+	root := t.rootV4
+	if isV6 {
+		root = t.rootV6
+	}
+	recursiveWalk(root, func(k netip.Prefix, v T) bool {
+		count++
+		return false
+	})
+	return count
+}
+
+// WalkFn is used when walking the tree. Takes a
+// key and value, returning if iteration should
+// be terminated.
+type WalkFn[T any] func(s netip.Prefix, v T) bool
+
+// DepthFirstWalk is used to walk the tree of the corresponding IP family
+func (t *Tree[T]) DepthFirstWalk(isIPv6 bool, fn WalkFn[T]) {
```

Comment:
I was going to do that but the v1.IPFamily is an enum with 3 values, this tree is IPv4 or IPv6, it can not be other thing, and create a yet another new enum just for this :(

Reply:
The third option being "" ?  Just return immediately - you found no IPs of family ""